In [1]:
import numpy as np
import pandas as pd

In [2]:
books = pd.read_csv("Books.csv")
users = pd.read_csv("Users.csv")
ratings = pd.read_csv("Ratings.csv")

C:\Users\Taha\AppData\Local\Temp\ipykernel_5940\2906776592.py:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv("Books.csv")


In [3]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [4]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [5]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [6]:
print(books.shape)
print(ratings.shape)
print(users.shape)


(271360, 8)
(1149780, 3)
(278858, 3)


In [7]:
books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [8]:
users.isnull().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

In [9]:
ratings.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

In [10]:
print(books.duplicated().sum())
print(users.duplicated().sum())
print(ratings.duplicated().sum())

0
0
0


### Popularity based recommender System:

In [11]:
ratings_with_name = ratings.merge(books, on = "ISBN")
ratings_with_name.head()

User-ID        ISBN  Book-Rating            Book-Title Book-Author  \
0   276725  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
1     2313  034545104X            5  Flesh Tones: A Novel  M. J. Rose   
2     6543  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
3     8680  034545104X            5  Flesh Tones: A Novel  M. J. Rose   
4    10314  034545104X            9  Flesh Tones: A Novel  M. J. Rose   

  Year-Of-Publication         Publisher  \
0                2002  Ballantine Books   
1                2002  Ballantine Books   
2                2002  Ballantine Books   
3                2002  Ballantine Books   
4                2002  Ballantine Books   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/034545104X.0...  
1  http://images.amazon.com/images/P/034545104X.0...  
2  http://images.amazon.com/images/P/034545104X.0...  
3  http://images.amazon.com/images/P/034545104X.0...  
4  http://images.amazon.com/images/P/034545104X.0...

In [12]:
ratings_with_name.groupby("Book-Title").count()

User-ID  ISBN  \
Book-Title                                                          
 A Light in the Storm: The Civil War Diary of A...        4     4   
 Always Have Popsicles                                    1     1   
 Apple Magic (The Collector's series)                     1     1   
 Ask Lily (Young Women of Faith: Lily Series, B...        1     1   
 Beyond IBM: Leadership Marketing and Finance f...        1     1   
...                                                     ...   ...   
Ã?Â?lpiraten.                                             2     2   
Ã?Â?rger mit Produkt X. Roman.                            4     4   
Ã?Â?sterlich leben.                                       1     1   
Ã?Â?stlich der Berge.                                     3     3   
Ã?Â?thique en toc                                         2     2   

                                                    Book-Rating  Book-Author  \
Book-Title                                                                     
 A Light in the Storm: The Civil War Diary of A...            4            4   
 Always Have Popsicles                                        1            1   
 Apple Magic (The Collector's series)                         1            1   
 Ask Lily (Young Women of Faith: Lily Series, B...            1            1   
 Beyond IBM: Leadership Marketing and Finance f...            1            1   
...                                                         ...          ...   
Ã?Â?lpiraten.                                                 2            2   
Ã?Â?rger mit Produkt X. Roman.                                4            4   
Ã?Â?sterlich leben.                                           1            1   
Ã?Â?stlich der Berge.                                         3            3   
Ã?Â?thique en toc                                             2            2   

                                                    Year-Of-Publication  \
Book-Title                                                                
 A Light in the Storm: The Civil War Diary of A...                    4   
 Always Have Popsicles                                                1   
 Apple Magic (The Collector's series)                                 1   
 Ask Lily (Young Women of Faith: Lily Series, B...                    1   
 Beyond IBM: Leadership Marketing and Finance f...                    1   
...                                                                 ...   
Ã?Â?lpiraten.                                                         2   
Ã?Â?rger mit Produkt X. Roman.                                        4   
Ã?Â?sterlich leben.                                                   1   
Ã?Â?stlich der Berge.                                                 3   
Ã?Â?thique en toc                                                     2   

                                                    Publisher  Image-URL-S  \
Book-Title                                                                   
 A Light in the Storm: The Civil War Diary of A...          4            4   
 Always Have Popsicles                                      1            1   
 Apple Magic (The Collector's series)                       1            1   
 Ask Lily (Young Women of Faith: Lily Series, B...          1            1   
 Beyond IBM: Leadership Marketing and Finance f...          1            1   
...                                                       ...          ...   
Ã?Â?lpiraten.                                               2            2   
Ã?Â?rger mit Produkt X. Roman.                              4            4   
Ã?Â?sterlich leben.                                         1            1   
Ã?Â?stlich der Berge.                                       3            3   
Ã?Â?thique en toc                                           2            2   

                                                    Image-URL-M  Image-URL-L  
Book-Title                                 

In [13]:
num_rating_df = ratings_with_name.groupby("Book-Title").count()["Book-Rating"].reset_index()
num_rating_df.rename(columns={"Book-Rating": "num_ratings"}, inplace = True)
num_rating_df

Book-Title  num_ratings
0        A Light in the Storm: The Civil War Diary of ...            4
1                                   Always Have Popsicles            1
2                    Apple Magic (The Collector's series)            1
3        Ask Lily (Young Women of Faith: Lily Series, ...            1
4        Beyond IBM: Leadership Marketing and Finance ...            1
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.            2
241067                     Ã?Â?rger mit Produkt X. Roman.            4
241068                                Ã?Â?sterlich leben.            1
241069                              Ã?Â?stlich der Berge.            3
241070                                  Ã?Â?thique en toc            2

[241071 rows x 2 columns]

In [14]:
num_rating_df["num_ratings"].max()

2502

In [15]:
avg_rating_df = ratings_with_name.groupby("Book-Title").mean()["Book-Rating"].reset_index()
avg_rating_df.rename(columns={"Book-Rating": "avg_ratings"}, inplace = True)
avg_rating_df

Book-Title  avg_ratings
0        A Light in the Storm: The Civil War Diary of ...     2.250000
1                                   Always Have Popsicles     0.000000
2                    Apple Magic (The Collector's series)     0.000000
3        Ask Lily (Young Women of Faith: Lily Series, ...     8.000000
4        Beyond IBM: Leadership Marketing and Finance ...     0.000000
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.     0.000000
241067                     Ã?Â?rger mit Produkt X. Roman.     5.250000
241068                                Ã?Â?sterlich leben.     7.000000
241069                              Ã?Â?stlich der Berge.     2.666667
241070                                  Ã?Â?thique en toc     4.000000

[241071 rows x 2 columns]

In [16]:
popular_df = num_rating_df.merge(avg_rating_df, on = "Book-Title")
popular_df

Book-Title  num_ratings  \
0        A Light in the Storm: The Civil War Diary of ...            4   
1                                   Always Have Popsicles            1   
2                    Apple Magic (The Collector's series)            1   
3        Ask Lily (Young Women of Faith: Lily Series, ...            1   
4        Beyond IBM: Leadership Marketing and Finance ...            1   
...                                                   ...          ...   
241066                                      Ã?Â?lpiraten.            2   
241067                     Ã?Â?rger mit Produkt X. Roman.            4   
241068                                Ã?Â?sterlich leben.            1   
241069                              Ã?Â?stlich der Berge.            3   
241070                                  Ã?Â?thique en toc            2   

        avg_ratings  
0          2.250000  
1          0.000000  
2          0.000000  
3          8.000000  
4          0.000000  
...             ...  
241066     0.000000  
241067     5.250000  
241068     7.000000  
241069     2.666667  
241070     4.000000  

[241071 rows x 3 columns]

In [17]:
popular_df = popular_df[popular_df["num_ratings"] >= 250].sort_values("avg_ratings", ascending = False).head(50)
popular_df

Book-Title  num_ratings  \
80434   Harry Potter and the Prisoner of Azkaban (Book 3)          428   
80422        Harry Potter and the Goblet of Fire (Book 4)          387   
80441      Harry Potter and the Sorcerer's Stone (Book 1)          278   
80426   Harry Potter and the Order of the Phoenix (Boo...          347   
80414    Harry Potter and the Chamber of Secrets (Book 2)          556   
191612  The Hobbit : The Enchanting Prelude to The Lor...          281   
187377  The Fellowship of the Ring (The Lord of the Ri...          368   
80445   Harry Potter and the Sorcerer's Stone (Harry P...          575   
211384     The Two Towers (The Lord of the Rings, Part 2)          260   
219741                              To Kill a Mockingbird          510   
183573                                  The Da Vinci Code          898   
187880                 The Five People You Meet in Heaven          430   
180556                             The Catcher in the Rye          449   
196326                          The Lovely Bones: A Novel         1295   
764                                                  1984          284   
144165                           Prodigal Summer: A Novel          253   
128670                                         Neverwhere          265   
206502                            The Secret Life of Bees          774   
168719  Stupid White Men ...and Other Sorry Excuses fo...          283   
223135  Tuesdays with Morrie: An Old Man, a Young Man,...          493   
204387                The Red Tent (Bestselling Backlist)          723   
191589               The Hitchhiker's Guide to the Galaxy          268   
129379   Nickel and Dimed: On (Not) Getting By in America          335   
93381                                       Into the Wild          252   
63867                                      Fahrenheit 451          409   
74750                           Girl with a Pearl Earring          526   
136145                                          Outlander          283   
233370  Where the Heart Is (Oprah's Book Club (Paperba...          585   
156102                     Seabiscuit: An American Legend          275   
107962                                         Life of Pi          664   
176845                                     The Bean Trees          389   
2281    A Child Called \It\": One Child's Courage to S...          265   
8434                                       ANGELA'S ASHES          279   
76343                                         Good in Bed          490   
64931   Fast Food Nation: The Dark Side of the All-Ame...          321   
12700                                       American Gods          302   
161645                                 Skipping Christmas          322   
105777  Left Behind: A Novel of the Earth's Last Days ...          318   
189551    The Golden Compass (His Dark Materials, Book 1)          336   
181679                                   The Color Purple          314   
160336                               Silence of the Lambs          256   
8752                                          About a Boy          262   
158138                  Seven Up (A Stephanie Plum Novel)          278   
175097  The Alchemist: A Fable About Following Your Dream          266   
80069   Hard Eight : A Stephanie Plum Novel (A Stephan...          269   
170101                       Suzanne's Diary for Nicholas          457   
111073                                  Lord of the Flies          259   
5664                              A Prayer for Owen Meany          413   
212070   The Vampire Lestat (Vampire Chronicles, Book II)          301   
233851       White Oleander : A Novel (Oprah's Book Club)          356   

        avg_ratings  
80434      5.852804  
80422      5.824289  
80441      5.737410  
80426      5.501441  
80414      5.183453  
191612     5.007117  
187377     4.948370  
80445      4.895652  
211384     4.880769  
219741     4.700000  
183573     4.642539  
187880  

>W e want to get the publisher name also

In [18]:
popular_df = popular_df.merge(books, on = "Book-Title").drop_duplicates("Book-Title")[["Book-Title", "Book-Author","Image-URL-M","num_ratings","avg_ratings"]]
popular_df

Book-Title  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)   
3         Harry Potter and the Goblet of Fire (Book 4)   
5       Harry Potter and the Sorcerer's Stone (Book 1)   
9    Harry Potter and the Order of the Phoenix (Boo...   
13    Harry Potter and the Chamber of Secrets (Book 2)   
16   The Hobbit : The Enchanting Prelude to The Lor...   
17   The Fellowship of the Ring (The Lord of the Ri...   
26   Harry Potter and the Sorcerer's Stone (Harry P...   
28      The Two Towers (The Lord of the Rings, Part 2)   
39                               To Kill a Mockingbird   
47                                   The Da Vinci Code   
53                  The Five People You Meet in Heaven   
55                              The Catcher in the Rye   
62                           The Lovely Bones: A Novel   
63                                                1984   
72                            Prodigal Summer: A Novel   
73                                          Neverwhere   
78                             The Secret Life of Bees   
84   Stupid White Men ...and Other Sorry Excuses fo...   
85   Tuesdays with Morrie: An Old Man, a Young Man,...   
88                 The Red Tent (Bestselling Backlist)   
89                The Hitchhiker's Guide to the Galaxy   
98    Nickel and Dimed: On (Not) Getting By in America   
100                                      Into the Wild   
103                                     Fahrenheit 451   
116                          Girl with a Pearl Earring   
117                                          Outlander   
123  Where the Heart Is (Oprah's Book Club (Paperba...   
124                     Seabiscuit: An American Legend   
126                                         Life of Pi   
130                                     The Bean Trees   
133  A Child Called \It\": One Child's Courage to S...   
135                                     ANGELA'S ASHES   
137                                        Good in Bed   
139  Fast Food Nation: The Dark Side of the All-Ame...   
140                                      American Gods   
141                                 Skipping Christmas   
147  Left Behind: A Novel of the Earth's Last Days ...   
149    The Golden Compass (His Dark Materials, Book 1)   
157                                   The Color Purple   
163                               Silence of the Lambs   
166                                        About a Boy   
173                  Seven Up (A Stephanie Plum Novel)   
175  The Alchemist: A Fable About Following Your Dream   
177  Hard Eight : A Stephanie Plum Novel (A Stephan...   
182                       Suzanne's Diary for Nicholas   
186                                  Lord of the Flies   
192                            A Prayer for Owen Meany   
194   The Vampire Lestat (Vampire Chronicles, Book II)   
195       White Oleander : A Novel (Oprah's Book Club)   

                Book-Author  \
0             J. K. Rowling   
3             J. K. Rowling   
5             J. K. Rowling   
9             J. K. Rowling   
13            J. K. Rowling   
16           J.R.R. TOLKIEN   
17           J.R.R. TOLKIEN   
26            J. K. Rowling   
28           J.R.R. TOLKIEN   
39               Harper Lee   
47                Dan Brown   
53              Mitch Albom   
55            J.D. Salinger   
62             Alice Sebold   
63            George Orwell   
72       Barbara Kingsolver   
73              Neil Gaiman   
78            Sue Monk Kidd   
84            Michael Moore   
85              MITCH ALBOM   
88            Anita Diamant   
89            Douglas Adams   
98       Barbara Ehrenreich   
100            Jon Krakauer   
103            Ray Bradbury   
116         Tracy Chevalier   
117          DIANA GABALDON   
123            Billie Letts   
124       LAURA HILLENBRAND   
126             Yann Martel   
130      Barbara Kingsolver   
133             Dave Pelzer   
135           Frank McCourt   
137         Jennifer Weiner   


In [19]:
popular_df["Image-URL-M"].values[0]

'http://images.amazon.com/images/P/0439136350.01.MZZZZZZZ.jpg'

>Collaborative Filtering:

In [20]:
ratings_with_name.head()

User-ID        ISBN  Book-Rating            Book-Title Book-Author  \
0   276725  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
1     2313  034545104X            5  Flesh Tones: A Novel  M. J. Rose   
2     6543  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
3     8680  034545104X            5  Flesh Tones: A Novel  M. J. Rose   
4    10314  034545104X            9  Flesh Tones: A Novel  M. J. Rose   

  Year-Of-Publication         Publisher  \
0                2002  Ballantine Books   
1                2002  Ballantine Books   
2                2002  Ballantine Books   
3                2002  Ballantine Books   
4                2002  Ballantine Books   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/034545104X.0...  
1  http://images.amazon.com/images/P/034545104X.0...  
2  http://images.amazon.com/images/P/034545104X.0...  
3  http://images.amazon.com/images/P/034545104X.0...  
4  http://images.amazon.com/images/P/034545104X.0...

In [21]:
x = ratings_with_name.groupby("User-ID").count()['Book-Rating']>200
regular_users = x[x].index
regular_users

Int64Index([   254,   2276,   2766,   2977,   3363,   4017,   4385,   6251,
              6323,   6543,
            ...
            271705, 273979, 274004, 274061, 274301, 274308, 275970, 277427,
            277639, 278418],
           dtype='int64', name='User-ID', length=811)

In [24]:
ratings_with_name

User-ID        ISBN  Book-Rating  \
0         276725  034545104X            0   
1           2313  034545104X            5   
2           6543  034545104X            0   
3           8680  034545104X            5   
4          10314  034545104X            9   
...          ...         ...          ...   
1031131   276688  0517145553            0   
1031132   276688  1575660792            7   
1031133   276690  0590907301            0   
1031134   276704  0679752714            0   
1031135   276704  0806917695            5   

                                                Book-Title      Book-Author  \
0                                     Flesh Tones: A Novel       M. J. Rose   
1                                     Flesh Tones: A Novel       M. J. Rose   
2                                     Flesh Tones: A Novel       M. J. Rose   
3                                     Flesh Tones: A Novel       M. J. Rose   
4                                     Flesh Tones: A Novel       M. J. Rose   
...                                                    ...              ...   
1031131                                    Mostly Harmless    Douglas Adams   
1031132                                        Gray Matter  Shirley Kennett   
1031133  Triplet Trouble and the Class Trip (Triplet Tr...     Debbie Dadey   
1031134  A Desert of Pure Feeling (Vintage Contemporaries)   Judith Freeman   
1031135  Perplexing Lateral Thinking Puzzles: Scholasti...      Paul Sloane   

        Year-Of-Publication                          Publisher  \
0                      2002                   Ballantine Books   
1                      2002                   Ballantine Books   
2                      2002                   Ballantine Books   
3                      2002                   Ballantine Books   
4                      2002                   Ballantine Books   
...                     ...                                ...   
1031131                1995             Random House Value Pub   
1031132                1996  Kensington Publishing Corporation   
1031133                1997                              Apple   
1031134                1997                  Vintage Books USA   
1031135                1997                Sterling Publishing   

                                               Image-URL-S  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/034545104X.0...   
2        http://images.amazon.com/images/P/034545104X.0...   
3        http://images.amazon.com/images/P/034545104X.0...   
4        http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1031131  http://images.amazon.com/images/P/0517145553.0...   
1031132  http://images.amazon.com/images/P/1575660792.0...   
1031133  http://images.amazon.com/images/P/0590907301.0...   
1031134  http://images.amazon.com/images/P/0679752714.0...   
1031135  http://images.amazon.com/images/P/0806917695.0...   

                                               Image-URL-M  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/034545104X.0...   
2        http://images.amazon.com/images/P/034545104X.0...   
3        http://images.amazon.com/images/P/034545104X.0...   
4        http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1031131  http://images.amazon.com/images/P/0517145553.0...   
1031132  http://images.amazon.com/images/P/1575660792.0...   
1031133  http://images.amazon.com/images/P/0590907301.0...   
1031134  http://images.amazon.com/images/P/0679752714.0...   
1031135  http://images.amazon.com/images/P/0806917695.0...   

                                               Image-URL-L  
0        http://images.amazon.com/images/P/034545104X.0...  
1        http://images.amazon.com/images/P/034545104X.0...  
2        http://images.amazon.com/images/P/034545104X

In [25]:
filtered_rating = ratings_with_name[ratings_with_name['User-ID'].isin(regular_users)]
filtered_rating

User-ID        ISBN  Book-Rating  \
2           6543  034545104X            0   
5          23768  034545104X            0   
7          28523  034545104X            0   
15         77940  034545104X            0   
16         81977  034545104X            0   
...          ...         ...          ...   
1030883   275970  1880837927            0   
1030884   275970  188717897X            0   
1030885   275970  1888889047            0   
1030886   275970  1931868123            0   
1030887   275970  3411086211           10   

                                                Book-Title  \
2                                     Flesh Tones: A Novel   
5                                     Flesh Tones: A Novel   
7                                     Flesh Tones: A Novel   
15                                    Flesh Tones: A Novel   
16                                    Flesh Tones: A Novel   
...                                                    ...   
1030883                         The Theology of the Hammer   
1030884  The Ordeal of Integration: Progress and Resent...   
1030885  Pushcart's Complete Rotten Reviews &amp; Rejec...   
1030886  There's a Porcupine in My Outhouse: Misadventu...   
1030887                                         Die Biene.   

                           Book-Author Year-Of-Publication  \
2                           M. J. Rose                2002   
5                           M. J. Rose                2002   
7                           M. J. Rose                2002   
15                          M. J. Rose                2002   
16                          M. J. Rose                2002   
...                                ...                 ...   
1030883                 Millard Fuller                1994   
1030884              Orlando Patterson                1998   
1030885                 Bill Henderson                1998   
1030886                   Mike Tougias                2002   
1030887  Sybil Gr�¤fin Sch�¶nfeldt                1993   

                                    Publisher  \
2                            Ballantine Books   
5                            Ballantine Books   
7                            Ballantine Books   
15                           Ballantine Books   
16                           Ballantine Books   
...                                       ...   
1030883         Smyth &amp; Helwys Publishing   
1030884                Civitas Book Publisher   
1030885                        Pushcart Press   
1030886                    Capital Books (VA)   
1030887  Bibliographisches Institut, Mannheim   

                                               Image-URL-S  \
2        http://images.amazon.com/images/P/034545104X.0...   
5        http://images.amazon.com/images/P/034545104X.0...   
7        http://images.amazon.com/images/P/034545104X.0...   
15       http://images.amazon.com/images/P/034545104X.0...   
16       http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1030883  http://images.amazon.com/images/P/1880837927.0...   
1030884  http://images.amazon.com/images/P/188717897X.0...   
1030885  http://images.amazon.com/images/P/1888889047.0...   
1030886  http://images.amazon.com/images/P/1931868123.0...   
1030887  http://images.amazon.com/images/P/3411086211.0...   

                                               Image-URL-M  \
2        http://images.amazon.com/images/P/034545104X.0...   
5        http://images.amazon.com/images/P/034545104X.0...   
7        http://images.amazon.com/images/P/034545104X.0...   
15       http://images.amazon.com/images/P/034545104X.0...   
16       http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1030883  http://images.amazon.com/images/P/1880837927.0...   
1030884  http://images.amazon.com/images/P/188717897X.0...   
1030885  http://images.amazon.com/images/P/1888889047.0...   
1030886  http://images.a

In [26]:
y = filtered_rating.groupby("Book-Title").count()['Book-Rating']>50
famous_books = y[y].index

In [27]:
final_ratings = filtered_rating[filtered_rating['Book-Title'].isin(famous_books)]
final_ratings

User-ID        ISBN  Book-Rating              Book-Title  \
63        278418  0446520802            0            The Notebook   
65          3363  0446520802            0            The Notebook   
66          7158  0446520802           10            The Notebook   
69         11676  0446520802           10            The Notebook   
74         23768  0446520802            6            The Notebook   
...          ...         ...          ...                     ...   
1026724   266865  0531001725           10  The Catcher in the Rye   
1027923   269566  0670809381            0                  Echoes   
1028777   271284  0440910927            0           The Rainmaker   
1029070   271705  B0001PIOX4            0          Fahrenheit 451   
1030868   275970  1586210661            9  Me Talk Pretty One Day   

                   Book-Author Year-Of-Publication                Publisher  \
63             Nicholas Sparks                1996             Warner Books   
65             Nicholas Sparks                1996             Warner Books   
66             Nicholas Sparks                1996             Warner Books   
69             Nicholas Sparks                1996             Warner Books   
74             Nicholas Sparks                1996             Warner Books   
...                        ...                 ...                      ...   
1026724  Jerome David Salinger                1973   Scholastic Library Pub   
1027923           Maeve Binchy                1986              Penguin USA   
1028777           John Grisham                1995                   Island   
1029070           Ray Bradbury                1993     Simon &amp; Schuster   
1030868          David Sedaris                2001  Time Warner Audio Major   

                                               Image-URL-S  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-M  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-L  
63       http://images.amazon.com/images/P/0446520802.0...  
65       http://images.amazon.com/images/P/0446520802.0...  
66       http://images.amazon.com/images/P/0446520802.0...  
69       http://images.amazon.com/images/P/0446520802.0...  
74       http://images.amazon.com/images/P/0446520802.0...  
...                                                    ...  
1026724  http://images.amazon.com/images/P/0531001725.0...  
1027923  http://images.amazon.com/images/P/0670809381.0...  
1028777  http://images.amazon.com/images/P/0440910927.0...  
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...  
1030868  http://images.amazon.com/images/P/1586210661.0...  

[57236 r

In [28]:
final_ratings["Image-URL-M"].values[0]

'http://images.amazon.com/images/P/0446520802.01.MZZZZZZZ.jpg'

In [29]:
pt = final_ratings.pivot_table(index = "Book-Title",columns = "User-ID", values = "Book-Rating")
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     0.0     NaN     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     0.0     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   NaN  ...    10.0   
1st to Die: A Novel                                    9.0  ...     NaN   
2nd Chance                                             0.0  ...     NaN   
4 Blondes                                              NaN  ...     NaN   
A Bend in the Road                                     NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

In [30]:
pt.fillna(0, inplace=True)

In [31]:
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

In [32]:
from sklearn.metrics.pairwise import cosine_similarity
similarity_scores = cosine_similarity(pt)
similarity_scores

array([[1.        , 0.10255025, 0.01220856, ..., 0.12110367, 0.07347567,
        0.04316046],
       [0.10255025, 1.        , 0.2364573 , ..., 0.07446129, 0.16773875,
        0.14263397],
       [0.01220856, 0.2364573 , 1.        , ..., 0.04558758, 0.04938579,
        0.10796119],
       ...,
       [0.12110367, 0.07446129, 0.04558758, ..., 1.        , 0.07085128,
        0.0196177 ],
       [0.07347567, 0.16773875, 0.04938579, ..., 0.07085128, 1.        ,
        0.10602962],
       [0.04316046, 0.14263397, 0.10796119, ..., 0.0196177 , 0.10602962,
        1.        ]])

In [33]:
cosine_similarity(pt).shape

(679, 679)

>Here i calculate the uclidean distance of 679 books with 679 books

In [34]:
similarity_scores[0]

array([1.        , 0.10255025, 0.01220856, 0.        , 0.05367224,
       0.02774901, 0.08216491, 0.13732869, 0.03261686, 0.03667591,
       0.02322418, 0.06766487, 0.02083978, 0.09673735, 0.08303112,
       0.11153543, 0.05100411, 0.02517784, 0.11706383, 0.        ,
       0.14333793, 0.07847534, 0.06150451, 0.08723968, 0.        ,
       0.07009814, 0.13658681, 0.07600328, 0.12167134, 0.00768046,
       0.01473221, 0.        , 0.07965814, 0.04522617, 0.01556271,
       0.09495938, 0.0182307 , 0.02610465, 0.07984012, 0.11679969,
       0.0569124 , 0.08354155, 0.08471898, 0.08785938, 0.05491435,
       0.0548505 , 0.27026514, 0.09779123, 0.06016046, 0.08958835,
       0.06748675, 0.        , 0.04468098, 0.01920872, 0.        ,
       0.05629067, 0.00557964, 0.07877059, 0.05219479, 0.18908177,
       0.        , 0.01240656, 0.02984572, 0.04279502, 0.12680125,
       0.16566735, 0.        , 0.13357242, 0.06615478, 0.        ,
       0.        , 0.        , 0.10968075, 0.02806606, 0.04521

In [35]:
np.where(pt.index == '1984')[0][0]

0

In [36]:
#0th index book will no be considered because it is the book itself
sorted(list(enumerate(similarity_scores[0])),key = lambda x:x[1],reverse = True)[1:6]

[(46, 0.2702651417103732),
 (524, 0.2639619371123497),
 (80, 0.23669374347400993),
 (610, 0.23299389358170394),
 (530, 0.2262639743141286)]

In [44]:
def recommend(book_name):
    index = np.where(pt.index == book_name)[0][0]
    similar_items = sorted(list(enumerate(similarity_scores[index])),key = lambda x:x[1],reverse = True)[1:6]
    data =[]
    for i in similar_items:
        #print(pt.index[i[0]])
        item = []
        tmp_df = books[books['Book-Title'] == pt.index[i[0]]]
        item.extend(list(tmp_df.drop_duplicates('Book-Title')['Book-Title'].values))
        item.extend(list(tmp_df.drop_duplicates('Book-Title')['Book-Author'].values))
        item.extend(list(tmp_df.drop_duplicates('Book-Title')['Image-URL-M'].values))
        data.append(item)
        
    return data
        

In [45]:
recommend("Harry Potter and the Chamber of Secrets (Book 2)")

[['Harry Potter and the Prisoner of Azkaban (Book 3)',
  'J. K. Rowling',
  'http://images.amazon.com/images/P/0439136350.01.MZZZZZZZ.jpg'],
 ['Harry Potter and the Goblet of Fire (Book 4)',
  'J. K. Rowling',
  'http://images.amazon.com/images/P/0439139597.01.MZZZZZZZ.jpg'],
 ["Harry Potter and the Sorcerer's Stone (Harry Potter (Paperback))",
  'J. K. Rowling',
  'http://images.amazon.com/images/P/059035342X.01.MZZZZZZZ.jpg'],
 ["Harry Potter and the Sorcerer's Stone (Book 1)",
  'J. K. Rowling',
  'http://images.amazon.com/images/P/0590353403.01.MZZZZZZZ.jpg'],
 ['Harry Potter and the Order of the Phoenix (Book 5)',
  'J. K. Rowling',
  'http://images.amazon.com/images/P/043935806X.01.MZZZZZZZ.jpg']]

In [101]:
import pickle
pickle.dump(popular_df, open("popular.pkl",'wb'))

In [47]:
import pickle
pickle.dump(pt, open("pt.pkl", "wb"))
pickle.dump(books, open("books.pkl", "wb"))
pickle.dump(similarity_scores, open("similarity_scores.pkl", "wb"))